File: src/main/java/com/hiberus/anaya/redmineeditor/utils/hiberus/Schedule.java
package com.hiberus.anaya.redmineeditor.utils.hiberus;

import javafx.scene.paint.Color;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.LocalDate;
import java.time.Month;
import java.time.YearMonth;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

/**
 * Logic regarding computed hours
 */
public class Schedule {

    /**
     * Calculates the hours you were supposed to spend on a day
     *
     * @param day which day to check
     * @return the hours you need to spend that day
     */
    static public double getExpectedHours(LocalDate day) {
        Month month = day.getMonth();
        int weekday = day.getDayOfWeek().getValue(); // 1 = monday, 7 = sunday

        // special days
        if (SPECIAL.containsKey(day)) return SPECIAL.get(day);

        if (month == Month.JULY || month == Month.AUGUST) {
            // summer schedule
            return new double[]{-1, 7, 7, 7, 7, 7, 0, 0}[weekday];
        } else {
            // normal schedule
            return new double[]{-1, 8.5, 8.5, 8.5, 8.5, 7, 0, 0}[weekday];
        }
    }

    /**
     * Calculates the hours you were supposed to spend on a month
     *
     * @param month which month to check
     * @return the hours you need to spend that month
     */
    public static double getExpectedHours(YearMonth month) {
        return IntStream
                // for each day of the month
                .rangeClosed(1, month.lengthOfMonth())
                // get the expected hours that day
                .mapToDouble(day -> getExpectedHours(month.atDay(day)))
                // and sum
                .sum();
    }

    /**
     * Calculates the color based on the day, and hours
     *
     * @param expected expected hours that day, probably from {@link Schedule#getExpectedHours(LocalDate)}
     * @param spent    spent hours that day
     * @param day      the day
     * @return the color of that day (null for no color)
     */
    static public Color getColor(double expected, double spent, LocalDate day) {
        if (expected != 0 && expected == spent) {
            // something to spend, and correctly spent, GOOD!
            return Color.LIGHTGREEN;
        } else if (expected == 0 && spent == 0) {
            // nothing to spend and nothing spent, HOLIDAY!
            return Color.LIGHTGREY;
        } else if (spent > expected) {
            // spent greater than expected, ERROR!
            return Color.RED;
        } else if (day.equals(LocalDate.now())) {
            // today, but still not all, WARNING!
            return Color.ORANGE;
        } else if (day.isBefore(LocalDate.now())) {
            // past day and not all, ERROR!
            return Color.RED;
        } else {
            // future day, NOTHING YET!
            return null; // null = no color
        }
    }


    /* ------------------------- Special days ------------------------- */
    private static final Map<LocalDate, Double> SPECIAL = new HashMap<>();

    static {
        // Load special days from hardcoded file
        String filename = "/home/anaya/abel/personal/proyectos/redmine/SpecialDays.csv";

        try (Stream<String> lines = Files.lines(Paths.get(filename))) {
            lines
                    .map(line -> line.replaceAll("#.*", "")) // remove comments
                    .filter(s -> !s.isBlank()) // skip empty
                    .forEach(line -> {
                        // parse data
                        List<String> data = Arrays.stream(line.split(",")).map(String::strip).collect(Collectors.toList());
                        if (data.size() < 3) {
                            // not enough
                            System.err.println("[ERROR] Not enough data in " + filename + "> " + line);
                        } else if (data.size() == 3) {
                            // year, month and day. 0 hours
                            SPECIAL.put(LocalDate.of(
                                    Integer.parseInt(data.get(0)),
                                    Integer.parseInt(data.get(1)),
                                    Integer.parseInt(data.get(2))
                            ), (double) 0);
                        } else {
                            // year, month, day and hours
                            SPECIAL.put(LocalDate.of(
                                    Integer.parseInt(data.get(0)),
                                    Integer.parseInt(data.get(1)),
                                    Integer.parseInt(data.get(2))
                            ), Double.parseDouble(data.get(3)));

                            if (data.size() > 4) {
                                // and other??
                                System.out.println("[Warning] more than necessary data in " + filename + "> " + line);
                            }
                        }
                    });
        } catch (IOException e) {
            e.printStackTrace();
            System.err.println("Special days file error!");
        }
    }

}

File: src/main/java/com/hiberus/anaya/redmineeditor/utils/hiberus/Schedule.kt
package com.hiberus.anaya.redmineeditor.utils.hiberus

import com.hiberus.anaya.redmineeditor.utils.findFile
import javafx.scene.paint.Color
import java.time.LocalDate
import java.time.Month
import java.time.YearMonth

/**
 * the expected hours you were supposed to spend this day
 */
val LocalDate.expectedHours
    // if special day, return it
    get() = SPECIAL.getOrElse(this) {
        // otherwise, get default
        when (month) {
            // summer schedule
            Month.JULY, Month.AUGUST -> doubleArrayOf(7.0, 7.0, 7.0, 7.0, 7.0, 0.0, 0.0)
            // normal schedule
            else -> doubleArrayOf(8.5, 8.5, 8.5, 8.5, 7.0, 0.0, 0.0)
        }[dayOfWeek.value - 1] // 0 = monday, 6 = sunday
    }

/**
 * the expected hours you were supposed to spend this month
 */
val YearMonth.expectedHours
    // just add all the hours of each day in month
    get() = (1..lengthOfMonth()).sumOf { atDay(it).expectedHours }

/**
 * Calculates the color based on the day, and hours
 *
 * @param expected expected hours that day, probably from [expectedHours]
 * @param spent    spent hours that day
 * @param day      the day
 * @return the color of that day (null for no color)
 */
fun getColor(expected: Double, spent: Double, day: LocalDate) = when {
    // something to spend, and correctly spent, GOOD!
    expected != 0.0 && expected == spent -> Color.LIGHTGREEN
    // nothing to spend and nothing spent, HOLIDAY!
    expected == 0.0 && spent == 0.0 -> Color.LIGHTGREY
    // spent greater than expected, ERROR!
    spent > expected -> Color.INDIANRED
    // today, but still not all, WARNING!
    day == LocalDate.now() -> Color.ORANGE
    // past day and not all, ERROR!
    day.isBefore(LocalDate.now()) -> Color.RED
    // future day, NOTHING! (null = no color)
    else -> null
}

/* ------------------------- Special days ------------------------- */

/**
 * true iff the special days were loaded
 * TODO: remove this
 */
var SpecialDaysLoaded = false

/**
 * Load special days from the configuration file
 */
fun LoadSpecialDays() = runCatching {
    SPECIAL.clear()
    SpecialDaysLoaded = false
    findFile("conf/special_days.conf").readLines().asSequence()
        // remove comments
        .map { it.replace("#.*".toRegex(), "") }
        // skip empty
        .filter { it.isNotBlank() }
        // split by comma
        .map { line -> line.split(",").map { it.trim() } to line }
        // build valid to map
        .mapNotNull { (data, line) ->
            when {
                data.size < 3 -> {
                    // not enough
                    System.err.println("Invalid entry in schedule, not enough data: $line")
                    null
                }
                data.size == 3 -> {
                    // year, month and day. 0 hours
                    LocalDate.of(data[0].toInt(), data[1].toInt(), data[2].toInt()) to 0.0
                }
                else -> {
                    if (data.size > 4) {
                        // and other??
                        println("Invalid entry in schedule, more than enough data: $line")
                    }

                    // year, month, day and hours
                    LocalDate.of(data[0].toInt(), data[1].toInt(), data[2].toInt()) to data[3].toDouble()
                }
            }
        }.toMap()
        // and save
        .let {
            SPECIAL.putAll(it)
            SpecialDaysLoaded = true
        }
}.onFailure {
    println(it)
    System.err.println("Special days file error!")
}.isSuccess

/* ------------------------- internal ------------------------- */

private val SPECIAL = mutableMapOf<LocalDate, Double>()